velocity/src/main/java/org/geysermc/floodgate/addon/data/VelocityServerDataHandler.java

```diff
@@ -119,7 +119,7 @@ public void write(ChannelHandlerContext ctx, Object packet, ChannelPromise promi
                 int addressFinished = address.indexOf('\0');
                 String originalAddress;
                 String remaining;
-                if (isModernForwarding && addressFinished == -1) {
+                if (isModernForwarding || addressFinished == -1) {
```

Comment:
I'm confused - why make this change?

Reply:
This `if` tries to prevent using `-1` in a call to `substring`, but with `&&` this is limited to only `isModernForwarding`, which is definitely wrong.

Changing this to '||' means to additionally force a different address format when Velocity with “modern” IP forwarding is used. It fixes error during connection (I will check again and write which exactly).

I verified that Velocity with “modern” IP forwarding is able to connect after this change. I have not checked what is really wrong with the packet format, but I can investigate it more.

I do not have Floodgate API running on the server, because the server is used also by Java clients at the same time and I do not need it. Anyway, I have more knowledge now, so I can investigate this more in depth. I have sources of Velocity, FastLogin and Floodgate, so I can check how it should really cooperate and if my fix is the right one.